text/3573-is.md

```diff
@@ -0,0 +1,270 @@
+- Feature Name: `is`
+- Start Date: 2024-02-16
+- RFC PR: [rust-lang/rfcs#3573](https://github.com/rust-lang/rfcs/pull/3573)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+
+Introduce an `is` operator in Rust 2024, to test if an expression matches a
+pattern and bind the variables in the pattern.
+
+# Motivation
+
+This RFC introduces an `is` operator that tests if an expression matches a
+pattern, and if so, binds the variables bound by the pattern and evaluates to
+true. This operator can be used as part of any boolean expression, and combined
+with boolean operators.
+
+Previous discussions around `let`-chains have treated the `is` operator as an
+alternative on the basis that they serve similar functions, rather than
+proposing that they can and should coexist. This RFC proposes that we allow
+`let`-chaining *and* add the `is` operator.
```

Comment:
I'm a bit concerned about this. This would introduce the possibility of doing the same thing in 2 different ways on a language level. IMHO this is a bad idea, as it opens the door for mixed-style code bases, that just get harder to read.

For tooling this is also a problem: Clippy will most likely get (restriction) lint requests for not allowing `is` OR not allowing `let`-chains. 

Another problem I see here is: What should Clippy do when producing suggestions? If we have the policy to always suggest `is` over `let`-chains, that might pollute code bases where `let`-chains are preferred (and vice versa). We also can't really check things like "is this a `let`-chain code base" or "are we in an `is`-chain expression" when producing suggestions. One lint suggesting `is` and another suggesting `let` will make this problem even worse, and that is almost impossible to avoid with changing contributors and team members.

We recently had the situation described above with suggesting the new-ish `_ =` binding over `let _ =`. We decided to suggest `let _ =` as we don't have to check the MSRV before producing the suggestion that way.

Reply:
Rust already has many different ways to do the same thing. You can write a `for` loop or you can write iterator code. You can use combinators or write a `match` or write an `if let`. You can write `let`-`else` or use a `match`. You can write `x > 3` or `3 < x`. You can write `x + 3` or `3 + x`.

> as it opens the door for mixed-style code bases, that just get harder to read

In this RFC, I'm proposing that both of them have value, and that it's entirely valid for a codebase to use both, for different purposes.

`if let PAT = EXPR && ...` emphasizes the pattern and its binding. It seems appropriate for clear division into cases based primarily on the pattern, by writing `if let ... else`.

`if EXPR is PAT && ...` leads with the expression, then the pattern, then the next condition. It feels more appropriate for cases where you expect the reader to find it easiest to process in order of the sequence of operations from left to right: "run this EXPR, see if it matches PAT, check the next condition ..."

I personally expect to find myself writing both, in different cases.

Reply:
For me those are not really comparable:
- The language gives you the `for` loop, the standard library gives you the option/power to do this with iterator method chains. 
- `match` is rather if you want to match one expression to multiple variants, while `if let` is for checking if the expression is that exact variant (there's a Clippy `style` lint for this).
- `let`-`else` was introduced for a specific use case to save some lines of code over using a `match`
- I hardly see how commutativity is related to this.

The proposed `is` language construct doesn't do the same:
- Both `let`-chains and `is` are provided as a language construct.
- There's no clear rule of thumb with `is` vs `let`. It's a pure style choice IMO.
- It doesn't simplify (in terms of amount of code) a certain, often occurring pattern.

The second point is the biggest problem for tooling: It is impossible to determine what to suggest. With the other examples it's usually clear, because the alternative is more concise/readable/idiomatic/.... 

---

The focus on expression vs pattern I can see and think is a valid point. But to that, I want to point out the [`equatable_if_let`](https://rust-lang.github.io/rust-clippy/master/index.html#/equatable_if_let?levels=allow) Clippy lint, that tried to address something similar, but never got out of `nursery` as we (mainly I) couldn't agree when `expr == pat` is preferable over `pat == expr`/`let pat = expr`. https://github.com/rust-lang/rust-clippy/pull/7777

So I see the addition of the `is` as giving the user a choice between two styles and not much more. IMO this is not worth the downsides that come with this. But that is my opinion and millage may vary obviously.

Reply:
I want to also link and quote one of my comments below: https://github.com/rust-lang/rfcs/pull/3573#issuecomment-1954267221 

> As `let`-chains are not stabilized yet, and iff there is consensus that the `is` approach is better, I think we should go with the `is` approach and remove `let`-chains again. I just think having both can cause problems and confusion, as I argued above.